File: 23b0931/w7_q3.py
MOD = 10**9 + 7

# Declare the DP array and the Trie data structure
dp = [0] * 5005
trie = [[0] * 26 for _ in range(10**6 + 5)]
isEndOfWord = [False] * (10**6 + 5)
trieNodeCount = 0

# Function to insert a word into the Trie
def insertWordInTrie(word):
    global trieNodeCount
    currentNode = 0
    for ch in word:
        index = ord(ch) - ord('a')
        if not trie[currentNode][index]:
            trieNodeCount += 1
            trie[currentNode][index] = trieNodeCount
        currentNode = trie[currentNode][index]
    isEndOfWord[currentNode] = True

# Function to count the number of ways to form the string
# from index 'start'
def countWays(start, S):
    currentNode = 0
    ways = 0
    for i in range(start, len(S)):
        index = ord(S[i]) - ord('a')
        if not trie[currentNode][index]:
            return ways
        currentNode = trie[currentNode][index]
        # If a word ends here, add the number of ways from the next index
        if isEndOfWord[currentNode]:
            ways = (ways + dp[i + 1]) % MOD
    return ways

def solve(S, K, word):
    for w in word:
        # Insert each word into the Trie
        insertWordInTrie(w)
    # Base case: One way to form an empty string
    dp[len(S)] = 1
    for i in range(len(S) - 1, -1, -1):
        # Fill the DP table from right to left
        dp[i] = countWays(i, S)
    # The answer is the number of ways to form the string from index 0
    print(dp[0])

# Driver Code
if __name__ == "__main__":
    # Input from user
    S = "ababc"
    K = 4
    arr = ["ab", "abab", "c", "cb"]
    solve(S, K, arr)

File: 23b1075/w7_q3.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll long long
const int mod = 1e9 + 7;

// Declare the DP array and the Trie data structure
vector<int> dp(5005);
vector<vector<int> > trie(1e6 + 5, vector<int>(26));
vector<bool> isEndOfWord(1e6 + 5);
int trieNodeCount = 0;

// Function to insert a word into the Trie
void insertWordInTrie(string& word)
{
    int currentNode = 0;
    for (char& ch : word) {
        if (!trie[currentNode][ch - 'a'])
            trie[currentNode][ch - 'a'] = ++trieNodeCount;
        currentNode = trie[currentNode][ch - 'a'];
    }
    isEndOfWord[currentNode] = true;
}

// Function to count the number of ways to form the string
// from index 'start'
int countWays(int start, string& S)
{
    int currentNode = 0, ways = 0;
    for (int i = start; i < S.size(); i++) {
        if (!trie[currentNode][S[i] - 'a'])
            return ways;
        currentNode = trie[currentNode][S[i] - 'a'];
        // If a word ends here, add the number of ways from the next index
        if (isEndOfWord[currentNode])
            ways = (ways + dp[i + 1]) % mod; 
    }
    return ways;
}

void solve(string& S, int K, string word[])
{
    for (int i = 0; i < K; i++) {
        // Insert each word into the Trie
        insertWordInTrie(word[i]); 
    }
    // Base case: One way to form an empty string
    dp[S.size()] = 1; 
    for (int i = S.size() - 1; i >= 0; i--) {
        // Fill the DP table from right to left
        dp[i] = countWays(i, S);
    }
    // The answer is the number of ways to form the string from index 0
    cout << dp[0]; 
}
// Driver Code
int main()
{
    // Input from user
    string S = "ababc";
    int K = 4;
    string arr[] = { "ab", "abab", "c", "cb" };
    solve(S, K, arr);
    return 0;
}
